File: ios/Runner/AssetDownload/AssetTransport.swift
import Foundation

/// Result of one ranged fetch performed by an `AssetTransport`.
struct AssetTransportChunk {
  /// Total size of the remote resource if known (from `Content-Length`/`Content-Range`).
  let totalSize: Int64?
  /// True when this response is known to cover the resource through EOF.
  let isComplete: Bool
  /// Number of bytes written to `destination` by this call.
  let bytesWritten: Int64
}

/// Abstracts the actual network transport so `AssetDownloadManager`/`AssetManifestFetcher`
/// are unit-testable without real network access, and so this framework stays reusable
/// for future consumers that might need a different transport (e.g. signed URLs).
protocol AssetTransport {
  /// Fetches bytes for `url` starting at byte offset `rangeStart`, appending them to
  /// the file at `destination` (creating it if needed; truncating first only if the
  /// server ignores the requested range and returns the full body from byte 0).
  func fetchAppending(url: URL, rangeStart: Int64, destination: URL) throws -> AssetTransportChunk

  /// Fetches a small JSON document in full (used for `catalog.json` / `model_manifest.json`).
  func fetchJSON(url: URL, timeout: TimeInterval) throws -> Data
}

/// Production transport: HTTP Range requests over `URLSession`. Presents a
/// synchronous API (via a semaphore) because every call site in this framework
/// already dispatches onto its own background queue — see `AssetDownloadManager`
/// and `TajweedAssetSync` — never the main thread.
///
/// Completions run on a `userInitiated` queue so a caller at that QoS is not
/// waiting on `URLSession.shared`'s default/utility threads (priority inversion).
final class URLSessionAssetTransport: AssetTransport {
  private let session: URLSession

  init(session: URLSession? = nil) {
    if let session {
      self.session = session
    } else {
      let queue = OperationQueue()
      queue.name = "com.rnr.deenfocus.assetdownload.urlsession"
      queue.qualityOfService = .userInitiated
      queue.maxConcurrentOperationCount = 4
      self.session = URLSession(configuration: .default, delegate: nil, delegateQueue: queue)
    }
  }

  func fetchJSON(url: URL, timeout: TimeInterval) throws -> Data {
    // Local file catalogs (DEBUG CoreML Official/DIY override) are not HTTP —
    // read them directly. Remote HTTPS catalogs keep the status-code check.
    if url.isFileURL {
      do {
        return try Data(contentsOf: url)
      } catch {
        throw AssetDownloadError.network(
          "Failed to read local catalog at \(url.path): \(error.localizedDescription)"
        )
      }
    }
    var request = URLRequest(url: url)
    request.timeoutInterval = timeout
    request.cachePolicy = .reloadIgnoringLocalCacheData
    let (data, response) = try syncDataTask(request)
    guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
      let code = (response as? HTTPURLResponse)?.statusCode ?? -1
      throw AssetDownloadError.badStatusCode(code)
    }
    return data
  }

  /// Bounded per-request chunk size. Artifacts can be several hundred MB (the
  /// DIY iOS CoreML encoder is ~590MB — see `memory/features/tajweed/`) — an
  /// open-ended `bytes=N-` Range request lets the server return the *entire*
  /// remaining object in one HTTP response, which can easily exceed
  /// `timeoutInterval` on a slow connection even though the transfer itself
  /// is still healthy. Requesting fixed-size chunks instead means each
  /// individual request comfortably fits the timeout regardless of total
  /// artifact size — `AssetDownloadManager`'s existing `while !isComplete`
  /// resume loop already handles multi-chunk transfers correctly; this was
  /// simply never exercised before because every response used to arrive as
  /// a single chunk.
  private static let chunkSizeBytes: Int64 = 8 * 1024 * 1024

  func fetchAppending(url: URL, rangeStart: Int64, destination: URL) throws -> AssetTransportChunk {
    var request = URLRequest(url: url)
    request.timeoutInterval = 30
    let rangeEnd = rangeStart + Self.chunkSizeBytes - 1
    request.setValue("bytes=\(rangeStart)-\(rangeEnd)", forHTTPHeaderField: "Range")
    let (data, response) = try syncDataTask(request)
    guard let http = response as? HTTPURLResponse else {
      throw AssetDownloadError.network("No HTTP response.")
    }
    guard (200..<300).contains(http.statusCode) else {
      throw AssetDownloadError.badStatusCode(http.statusCode)
    }

    let fm = FileManager.default
    if http.statusCode == 200, rangeStart > 0 {
      // Server ignored our Range header; response is the full file from byte 0.
      try? fm.removeItem(at: destination)
    }
    if !fm.fileExists(atPath: destination.path) {
      fm.createFile(atPath: destination.path, contents: nil)
    }
    let handle = try FileHandle(forWritingTo: destination)
    defer { try? handle.close() }
    handle.seekToEndOfFile()
    handle.write(data)

    let totalSize: Int64? = {
      if let contentRange = http.value(forHTTPHeaderField: "Content-Range"),
        let slashIdx = contentRange.lastIndex(of: "/")
      {
        return Int64(contentRange[contentRange.index(after: slashIdx)...])
      }
      if http.expectedContentLength > 0 {
        return http.expectedContentLength + (http.statusCode == 206 ? rangeStart : 0)
      }
      return nil
    }()

    let newSize = (try? fm.attributesOfItem(atPath: destination.path)[.size] as? Int64) ?? nil
    let isComplete: Bool
    if let totalSize, let newSize {
      isComplete = newSize >= totalSize
    } else {
      // No length info from the server: a single successful fetch is treated as
      // complete (matches static object storage, which always returns the full
      // remaining body per request rather than arbitrary partial chunks).
      isComplete = true
    }
    return AssetTransportChunk(totalSize: totalSize, isComplete: isComplete, bytesWritten: Int64(data.count))
  }

  private func syncDataTask(_ request: URLRequest) throws -> (Data, URLResponse) {
    let semaphore = DispatchSemaphore(value: 0)
    var result: Result<(Data, URLResponse), Error> = .failure(AssetDownloadError.timeout)
    let task = session.dataTask(with: request) { data, response, error in
      if let error {
        result = .failure(AssetDownloadError.network(error.localizedDescription))
      } else if let data, let response {
        result = .success((data, response))
      } else {
        result = .failure(AssetDownloadError.network("Empty response."))
      }
      semaphore.signal()
    }
    task.resume()
    let waitResult = semaphore.wait(timeout: .now() + request.timeoutInterval + 5)
    if waitResult == .timedOut {
      task.cancel()
      throw AssetDownloadError.timeout
    }
    return try result.get()
  }
}

File: android/app/src/main/kotlin/com/app/deenly/deenly/assetdownload/AssetTransport.kt
package com.rnr.deenfocus.assetdownload

import android.util.Log
import java.io.RandomAccessFile
import java.net.HttpURLConnection
import java.net.URL

private const val TAG = "AssetTransport"

/**
 * Max bytes read per HTTP round-trip in [HttpUrlConnectionAssetTransport.fetchAppending].
 * Bounds peak memory to a small, fixed multiple of this value (via the streaming copy
 * loop's own 64KB buffer) regardless of how large the remote asset is — critical for
 * ~450MB+ model packs on memory-constrained devices/emulators. Also gives
 * [AssetDownloadManager]'s outer loop a natural, frequent resume point if the process
 * dies mid-file, and smoother progress reporting than one giant single-shot read.
 */
private const val DOWNLOAD_CHUNK_BYTES = 8L * 1024 * 1024

/** Result of one ranged fetch performed by an [AssetTransport]. */
data class AssetTransportChunk(
    /** Total size of the remote resource if known (from Content-Length/Content-Range). */
    val totalSize: Long?,
    /** True when this response is known to cover the resource through EOF. */
    val isComplete: Boolean,
    /** Number of bytes written to `destinationPath` by this call. */
    val bytesWritten: Long,
)

/**
 * Abstracts the actual network transport so [AssetDownloadManager]/[AssetManifestFetcher]
 * are unit-testable without real network access, and so this framework stays reusable
 * for future consumers that might need a different transport (e.g. signed URLs).
 */
interface AssetTransport {
    /**
     * Fetches bytes for [url] starting at byte offset [rangeStart], appending them to
     * the file at [destinationPath] (creating it if needed; truncating first only if
     * the server ignores the requested range and returns the full body from byte 0).
     */
    fun fetchAppending(url: String, rangeStart: Long, destinationPath: java.io.File): AssetTransportChunk

    /** Fetches a small JSON document in full (used for catalog.json / model_manifest.json). */
    fun fetchJson(url: String, timeoutMs: Int): String
}

/**
 * Production transport: HTTP Range requests over [HttpURLConnection] (no new
 * dependency — no OkHttp/Retrofit needed, consistent with this codebase's
 * dependency-light convention, e.g. TD-008's hand-rolled FFT). Synchronous by
 * design; callers (`AssetDownloadManager`) always run this off the main thread.
 */
class HttpUrlConnectionAssetTransport : AssetTransport {

    override fun fetchJson(url: String, timeoutMs: Int): String {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.connectTimeout = timeoutMs
        connection.readTimeout = timeoutMs
        connection.requestMethod = "GET"
        connection.setRequestProperty("Cache-Control", "no-cache")
        try {
            val code = connection.responseCode
            Log.d(TAG, "[http] GET $url -> $code")
            if (code !in 200..299) {
                throw AssetDownloadException.BadStatusCode(code)
            }
            return connection.inputStream.use { it.readBytes() }.toString(Charsets.UTF_8)
        } catch (e: java.io.IOException) {
            Log.w(TAG, "[http] GET $url failed: ${e.message}")
            throw AssetDownloadException.Network(e.message ?: "network error")
        } finally {
            connection.disconnect()
        }
    }

    override fun fetchAppending(url: String, rangeStart: Long, destinationPath: java.io.File): AssetTransportChunk {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.connectTimeout = 30_000
        connection.readTimeout = 30_000
        connection.requestMethod = "GET"
        if (rangeStart > 0) {
            connection.setRequestProperty("Range", "bytes=$rangeStart-")
        }
        try {
            val code = connection.responseCode
            Log.d(TAG, "[http] GET $url (Range: bytes=$rangeStart-) -> $code")
            if (code !in 200..299) {
                throw AssetDownloadException.BadStatusCode(code)
            }
            // Server ignored our Range header; response is the full file from byte 0.
            if (code == 200 && rangeStart > 0) {
                Log.w(TAG, "[http] Server ignored Range header for $url — restarting $destinationPath from byte 0.")
                destinationPath.delete()
            }

            val contentRange = connection.getHeaderField("Content-Range")
            val totalSize: Long? = when {
                contentRange != null && contentRange.contains("/") ->
                    contentRange.substringAfterLast("/").toLongOrNull()
                connection.contentLengthLong > 0 ->
                    connection.contentLengthLong + (if (code == 206) rangeStart else 0)
                else -> null
            }

            // Parent must exist before RandomAccessFile — otherwise Android throws
            // ENOENT ("…onnx.part: open failed: ENOENT") which surfaces as a
            // confusing network error instead of a disk/setup failure.
            ensureParentDir(destinationPath)

            // Stream at most DOWNLOAD_CHUNK_BYTES to disk via a small fixed buffer —
            // never buffer the whole (potentially 450MB+) response body in memory.
            // If more than one chunk remains, isComplete=false and the caller
            // (AssetDownloadManager.downloadOne) issues a fresh Range request
            // continuing from the new offset.
            var written = 0L
            var reachedEof = false
            try {
                RandomAccessFile(destinationPath, "rw").use { raf ->
                    raf.seek(raf.length())
                    connection.inputStream.use { input ->
                        val buffer = ByteArray(64 * 1024)
                        while (written < DOWNLOAD_CHUNK_BYTES) {
                            val toRead = minOf(buffer.size.toLong(), DOWNLOAD_CHUNK_BYTES - written).toInt()
                            val n = input.read(buffer, 0, toRead)
                            if (n <= 0) {
                                reachedEof = true
                                break
                            }
                            raf.write(buffer, 0, n)
                            written += n
                        }
                    }
                }
            } catch (e: java.io.IOException) {
                throw mapIoException(e, destinationPath)
            }

            val newSize = destinationPath.length()
            val isComplete = if (totalSize != null) newSize >= totalSize else reachedEof
            return AssetTransportChunk(totalSize = totalSize, isComplete = isComplete, bytesWritten = written)
        } catch (e: AssetDownloadException) {
            throw e
        } catch (e: java.io.IOException) {
            Log.w(TAG, "[http] GET $url (Range: bytes=$rangeStart-) failed: ${e.message}")
            throw mapIoException(e, destinationPath)
        } finally {
            connection.disconnect()
        }
    }

    private fun ensureParentDir(file: java.io.File) {
        val parent = file.parentFile ?: return
        if (parent.exists()) return
        if (!parent.mkdirs() && !parent.exists()) {
            val available = parent.parentFile?.usableSpace ?: 0L
            throw AssetDownloadException.DiskError(
                "Could not create download directory ${parent.absolutePath} " +
                    "(available=${available} bytes). Free space and retry.",
            )
        }
    }

    private fun mapIoException(e: java.io.IOException, destinationPath: java.io.File): AssetDownloadException {
        val msg = e.message ?: "I/O error"
        Log.w(TAG, "[http] write to ${destinationPath.absolutePath} failed: $msg")
        return when {
            msg.contains("ENOSPC", ignoreCase = true) ||
                msg.contains("No space left", ignoreCase = true) -> {
                val available = destinationPath.parentFile?.usableSpace ?: 0L
                AssetDownloadException.InsufficientStorage(
                    requiredBytes = destinationPath.length().coerceAtLeast(1L),
                    availableBytes = available,
                )
            }
            msg.contains("ENOENT", ignoreCase = true) ->
                AssetDownloadException.DiskError(msg)
            else -> AssetDownloadException.Network(msg)
        }
    }
}
